runtime/lua/vim/lsp/sync.lua

```diff
@@ -93,31 +93,38 @@ end
 -- utf-8 index and either the utf-16, or utf-32 index.
 ---@param line string the line to index into
 ---@param byte integer the byte idx
+---@param start boolean true for start align, false for end align
 ---@param offset_encoding string utf-8|utf-16|utf-32|nil (default: utf-8)
 ---@returns table<string, int> byte_idx and char_idx of first change position
-local function align_end_position(line, byte, offset_encoding)
-  local char
-  -- If on the first byte, or an empty string: the trivial case
-  if byte == 1 or #line == 0 then
-    char = byte
-  -- Called in the case of extending an empty line "" -> "a"
-  elseif byte == #line + 1 then
-    char = compute_line_length(line, offset_encoding) + 1
-  else
+local function align_position(line, byte, start, offset_encoding)
+  if byte ~= 1 and byte <= #line then
     -- Modifying line, find the nearest utf codepoint
     local offset = str_utf_start(line, byte)
+
     -- If the byte does not fall on the start of the character, then
-    -- align to the start of the next character.
+    -- align to the start of the next character if end align, and start
+    -- of this character if start align
     if offset < 0 then
-      byte = byte + str_utf_end(line, byte) + 1
-    end
-    if byte <= #line then
-      char = byte_to_utf(line, byte, offset_encoding)
-    else
-      char = compute_line_length(line, offset_encoding) + 1
+      if start then
+        byte = byte + offset
+      else
+        byte = byte + str_utf_end(line, byte) + 1
+      end
     end
-    -- Extending line, find the nearest utf codepoint for the last valid character
   end
+
+  local char
+
+  -- optimize for first byte case
+  if byte == 1 then
+    char = 1
+  -- Called in the case of extending an empty line "" -> "a"
+  elseif byte == #line + 1 then
+    char = compute_line_length(line, offset_encoding) + 1
```

Comment:
These functions can be easily unified since I allowed byteindex to take explicit nuls

Reply:
See #16731, we should be able to just use those helpers directly.

Reply:
^ done, let me know what you think